reactjs-todo-dv/client/components/davinci-client/davinci-flow.js

```diff
@@ -0,0 +1,180 @@
+    setComponents([errorMessage]);
+  };
+
+  // Represents the main render function for app
+  async function renderForm(nextNode) {
+    const arrFormFields = [];
+    // clear form contents
+    setComponents([]);
+    // Set h1 header
+    setPageHeader(nextNode.client?.name || '');
+
+    const collectors = client.current.collectors();
+
+    collectors.forEach((collector) => {
+      if (collector.type === 'TextCollector' && collector.name === 'protectsdk') {
+        arrFormFields.push(
+          <Protect
+            collector={collector}
+            updater={client.current.update(collector)}
+            key={`protect-${collector.output.key}`}
+          />,
+        );
+      } else if (collector.type === 'TextCollector') {
+        arrFormFields.push(
+          <TextInput
+            collector={collector}
+            updater={client.current.update(collector)}
+            key={`text-${collector.output.key}`}
+          />,
+        );
+      } else if (collector.type === 'PasswordCollector') {
+        arrFormFields.push(
+          <Password
+            collector={collector}
+            updater={client.current.update(collector)}
+            key={`password-${collector.output.key}`}
+          />,
+        );
+      } else if (collector.type === 'SubmitCollector') {
+        arrFormFields.push(
+          <SubmitButton collector={collector} key={`submit-btn-${collector.output.key}`} />,
+        );
+      } else if (collector.type === 'FlowCollector') {
+        arrFormFields.push(
+          <FlowButton
+            collector={collector}
+            key={`flow-btn-${collector.output.key}`}
+            flow={client.current.flow({ action: collector.output.key })}
+            renderForm={renderForm}
+          />,
+        );
+      }
+    });
+
+    // Save components to state
+    setComponents(arrFormFields);
+
+    if (client.current.collectors().find((collector) => collector.name === 'protectsdk')) {
+      const newNode = await client.current.next();
+
+      if (newNode.status === 'next') {
+        return renderForm(newNode);
+      } else if (newNode.status === 'success') {
+        return renderComplete(newNode);
+      } else if (newNode.status === 'error') {
+        return renderError(newNode);
+      } else {
+        console.error('Unknown node status', newNode);
+      }
+    }
+  }
+
+  const onSubmitHandler = async (event) => {
+    event.preventDefault();
+
+    /**
+     * We can just call `next` here and not worry about passing any arguments
+     */
+    const newNode = await client.current.next();
+
+    /**
+     * Recursively render the form with the new state
+     */
+    if (newNode.status === 'next') {
+      renderForm(newNode);
+    } else if (newNode.status === 'success') {
+      return renderComplete(newNode);
+    } else if (newNode.status === 'error') {
+      return renderError(newNode);
+    } else {
+      console.error('Unknown node status', newNode);
+    }
```

Comment:
To avoid having to redeclare this conditional logic multiple times, let's move this down into the return statement. To ensure this component rerenders, we use the fact that the `async-await` function above completes and new state is introduced to the component, either through a `setState` or from a custom React Hook, and that's what triggers a rerender and this conditional logic is rerun in the return statement.